### PR TITLE
gh-102560 Add docstrings to asyncio.TaskGroup

### PR DESCRIPTION
**Pull request by JosephSBoyle**
Adds docstrings to the public methods of asyncio.TaskGroup.

* Issue: gh-102560

### CODE REVIEW

**Review by AlexWaygood — DISMISSED** [review]
Please fix the trailing whitespace issues that the Azure Pipelines CI check is complaining about

File: Lib/asyncio/taskgroups.py [hunk 1]
```diff
@@ -135,6 +158,10 @@ async def __aexit__(self, et, exc, tb):
                 self._errors = None
 
     def create_task(self, coro, *, name=None, context=None):
+        """Create a new task in this group and return it.
+        
+        Matches the call signature of asyncio.create_task.
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py]: Why not just
```suggestion
        Similar to `asyncio.create_task`.
```

**Review by gvanrossum — COMMENTED** [review]
I desperate wish for something half the size.

**Comment by gvanrossum** [comment]
@JosephSBoyle Can you please respond to the review comments? (PS. There's no need to merge main, even if GitHub recommends it, _unless_ it says there's also a merge conflict.)

**Comment by JosephSBoyle** [comment]
Thanks for the comments!

@AlexWaygood I believe I've fixed the whitespace issue.
@gvanrossum and @olijeffers0n I've adopted your suggestions.

Thanks for the tip on merging! :bowing_man: 

---------------------------------
A thought of my own, maybe we could remove the 'Similar to `asyncio.create_task' line entirely and just have:
```python3
    def create_task(self, coro, *, name=None, context=None):
        """Create a new task in this group and return it."""

```
I think people intuitively would expect `asyncio.create_task` and `asyncio.TaskGroup().create_task` to be similar from name alone which makes the statement redundant.

File: Lib/asyncio/taskgroups.py [hunk 2]
```diff
@@ -10,7 +10,22 @@
...
+
+        async with asyncio.TaskGroup() as group:
+            task1 = group.create_task(some_coroutine(...))
+            task2 = group.create_task(other_coroutine(...))
+
+        print("Both tasks have completed now.")
+
+    All tasks are awaited when the context manager exits.
+
+    Any exceptions other than asyncio.CancelledError's raised within
```
Comment (AlexWaygood) [inline Lib/asyncio/taskgroups.py]: The grammar here isn't quite right. Maybe just:

```suggestion
    Any exceptions other than asyncio.CancelledError raised within
```

**Comment by kumaraditya303** [comment]
I defer to Alex and Guido on this.

**Comment by AlexWaygood** [comment]
It reads okay to me now, but neither an `asyncio` maintainer nor an `asyncio` expert, so I'll defer to @gvanrossum and Kumar on whether the technical accuracy is okay and whether the style of the docstring is desirable.

Thanks for working on this!

**Comment by JosephSBoyle** [comment]
Thanks for your help Alex:)
:bowing_man:

File: Lib/asyncio/taskgroups.py [hunk 3]
```diff
@@ -10,7 +10,22 @@
 
 
 class TaskGroup:
+    """An asynchronous context manager for managing groups of tasks.
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py]:
```suggestion
    """Asynchronous context manager for managing groups of tasks.
```

File: Lib/asyncio/taskgroups.py [hunk 4]
```diff
@@ -10,7 +10,22 @@
...
 
+    Example use:
+
+        async with asyncio.TaskGroup() as group:
+            task1 = group.create_task(some_coroutine(...))
+            task2 = group.create_task(other_coroutine(...))
+
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py]: We can do without this blank line.
```suggestion
```

File: Lib/asyncio/taskgroups.py [hunk 5]
```diff
@@ -10,7 +10,22 @@
...
+        print("Both tasks have completed now.")
+
+    All tasks are awaited when the context manager exits.
+
+    Any exceptions other than asyncio.CancelledErrors raised within
+    tasks will cancel all remaining tasks and exit the context. These
+    exceptions will be combined into an ExceptionGroup and re-raised.
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py]:
```suggestion
    Any exceptions other than `asyncio.CancelledError` raised within
    a task will cancel all remaining tasks and wait for them to exit.
    The exceptions are then combined and raised as an `ExceptionGroup`.
```

File: Lib/asyncio/taskgroups.py [hunk 6]
```diff
@@ -135,6 +150,9 @@ async def __aexit__(self, et, exc, tb):
                 self._errors = None
 
     def create_task(self, coro, *, name=None, context=None):
+        """Create a new task in this group and return it.
+        Similar to `asyncio.create_task`.
```
Comment (gvanrossum) [inline Lib/asyncio/taskgroups.py]: Multi-line docstrings should have a blank line after the first line.
```suggestion
        """Create a new task in this group and return it.

        Similar to `asyncio.create_task`.
```

**Review by gvanrossum — COMMENTED** [review]
I have a slight rephrasing and some whitespace/markup nits, then LGTM.

I can commit those and land this  unless you object.

**Comment by JosephSBoyle** [comment]
Hi @gvanrossum, feel free to commit the changes you proposed:)

**Comment by gvanrossum** [comment]
Thanks! It's a real improvement to have this.

**Comment by JosephSBoyle** [comment]
Happy to help!:)
🙇‍♂️

**Comment by gvanrossum** [comment]
If you want another project like this, timeouts.py also needs docstrings.

**Comment by JosephSBoyle** [comment]
> If you want another project like this, timeouts.py also needs docstrings.

Nice, I've opened a PR to add docstrings here: https://github.com/python/cpython/pull/102811.

**Pull request merged**